agent/handlers/agentapi/taskprotection/v1/handlers/handlers.go

```diff
@@ -19,75 +19,186 @@ import (
+		})
+
+		logger.Debug("updateTaskProtection response:", logger.Fields{
+			loggerfield.TaskProtection: response.ProtectedTasks,
+			loggerfield.Error:          err,
+			loggerfield.Reason:         response.Failures,
+		})
+
+		if err != nil {
+			exceptionType, statusCode := getExceptionTypeAndStatusCode(err)
+			logger.Info(fmt.Sprintf("ECS throws an exception with type %s", exceptionType))
+			writeJSONResponse(w, statusCode, err.Error(), updateTaskProtectionRequestType)
+			return
+		}
+
+		// there are no exceptions but there are failures when setting protection in scheduler
+		if len(response.Failures) > 0 {
+			writeJSONResponse(w, http.StatusBadRequest, response.Failures[0], updateTaskProtectionRequestType)
+			return
+		}
+		if len(response.ProtectedTasks) != ExpectedProtectionResponseLength {
+			msg := fmt.Sprintf("expect %v protectedTask in response, get %v", ExpectedProtectionResponseLength, len(response.ProtectedTasks))
+			logger.Error(msg)
+			writeJSONResponse(w, http.StatusInternalServerError, msg, updateTaskProtectionRequestType)
+			return
+		}
+		writeJSONResponse(w, http.StatusOK, response.ProtectedTasks[0], updateTaskProtectionRequestType)
+	}
+}
+
+// Helper function for retrieving credential from credentials manager and create ecs client
+func (factory TaskProtectionClientFactory) getECSClient(credentialsManager credentials.Manager, task *apitask.Task) (api.ECSTaskProtectionSDK, int, error) {
+	taskRoleCredential, ok := credentialsManager.GetTaskCredentials(task.GetCredentialsID())
+	if !ok {
+		return nil, http.StatusUnauthorized, errors.New("invalid Request: no task IAM role credentials available for task")
+	}
+	taskCredential := taskRoleCredential.GetIAMRoleCredentials()
+	cfg := aws.NewConfig().
+		WithCredentials(awscreds.NewStaticCredentials(taskCredential.AccessKeyID,
+			taskCredential.SecretAccessKey,
+			taskCredential.SessionToken)).
+		WithRegion(factory.Region).
+		WithHTTPClient(httpclient.New(ecsclient.RoundtripTimeout, factory.AcceptInsecureCert)).
+		WithEndpoint(factory.Endpoint)
+
+	ecsClient := ecs.New(session.Must(session.NewSession()), cfg)
+	return ecsClient, http.StatusOK, nil
+}
+
+// Helper function to parse exception type from error since it is unmarshalled in ecs client
+// see newClient method in agent/ecs_client/model/ecs/service.go
+func getExceptionTypeAndStatusCode(err error) (string, int) {
+	errorType := ""
+	// typical error format: InvalidParameterException: <Exception Message>
+	for i := range err.Error() {
+		if err.Error()[i] == ' ' {
+			errorType = err.Error()[0 : i-1] // trim colon
+			break
+		}
+	}
+
+	switch errorType {
+	case ecs.ErrCodeServerException:
+		return errorType, http.StatusInternalServerError
+	case ecs.ErrCodeAccessDeniedException,
+		ecs.ErrCodeClientException,
+		ecs.ErrCodeClusterNotFoundException,
+		ecs.ErrCodeInvalidParameterException,
+		ecs.ErrCodeResourceNotFoundException,
+		ecs.ErrCodeUnsupportedFeatureException:
```

Comment:
I think all of these errors have corresponding types in the AWS SDK. We should be able to use standard Go error handling [as described here](https://aws.github.io/aws-sdk-go-v2/docs/handling-errors/) to get the underlying error. Did you try this?

For example, AccessDeniedException is defined [here](https://github.com/aws/amazon-ecs-agent/blob/master/agent/ecs_client/model/ecs/api.go#L3851).

Reply:
Unfortunately our repo seems to use aws-sdk-go v1 instead of aws-sdk-go-v2, and  I can’t type assert error into types defined in api.go. Considering the effort of creating a wrapper (as in acs) or adopt to v2 go sdk, I didn't do it.

Reply:
Looking at v1 sdk https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/handling-errors.html as suggested by @amogh09

Reply:
updated.